batch/batch/server/dag.py

```diff
@@ -0,0 +1,130 @@
+from .. import data
+
+from .globals import dag_id_dag
+from .job import Job
+from .log import log
+from .user_error import UserError
+
+
+class DagNode:
+    def __init__(self, spec, parents, children):
+        self.spec = spec
+        self._parents = parents
+        self.children = children
+        self.unfinished_parents = set([])
+        self.job = None
+
+    def add_parent(self, parent):
+        self._parents.append(parent)
+        self.unfinished_parents.add(parent.spec.name)
+
+    def add_child(self, child):
+        self.children.append(child)
+
+    def link_job(self, job):
+        assert self.job is None
+        self.job = job
+
+    def to_json(self):
+        return {
+            'spec': self.spec.to_json(),
+            'parents': [parent.spec.name for parent in self._parents],
+            'children': [child.spec.name for child in self.children]
+        }
+
+    def to_get_json(self):
+        return {
+            'job': self.job,
+            'parents': [parent.spec.name for parent in self._parents],
+            'children': [child.spec.name for child in self.children]
+        }
+
+    def __str__(self):
+        return str(self.to_json())
+
+    def __repr__(self):
+        return self.__str__()
+
+    def mark_complete(self):
+        exit_code = self.job.exit_code
+        if exit_code == 0:
+            for child in self.children:
+                child.mark_parent_complete(self)
+        else:
+            log.warning(f'failing pod {exit_code} will not trigger dag completion')
+
+    def mark_parent_complete(self, parent):
+        parent_name = parent.spec.name
+        log.info(f'parent {parent_name}/{parent.job.id} completed for child {self.spec.name}')
+        if parent_name not in self.unfinished_parents:
+            raise ValueError(
+                f'unknown parent: {parent_name} not in {self.unfinished_parents}. '
+                f'{self}')
+        self.unfinished_parents.remove(parent_name)
+        if not self.unfinished_parents:
+            log.info(f'all parents completed for child {self.spec.name}')
+            assert self.job is None
+            self.create_job()
+            assert self.job is not None
+
+    def create_job(self):
+        return Job(self.spec.job_spec, self)
+
+    def delete(self):
+        if self.job:
+            self.job.delete()
+
+    def cancel(self):
+        if self.job:
+            self.job.cancel()
+
+
+class Dag:
```

Comment:
I'm a bit uncomfortable with class name overloading. I think this'll cause problems in the future. Maybe we could give the data proxy classes some suffix to clarify?